plugins/outputs/opensearch/opensearch.go

```diff
@@ -0,0 +1,451 @@
+	var buffer bytes.Buffer
+	buffer.WriteString(strconv.FormatInt(m.Time().Local().UnixNano(), 10))
+	buffer.WriteString(m.Name())
+	buffer.WriteString(strconv.FormatUint(m.HashID(), 10))
+
+	return fmt.Sprintf("%x", sha256.Sum256(buffer.Bytes()))
+}
+
+func (o *Opensearch) Write(metrics []telegraf.Metric) error {
+	// get indexers based on unique pipeline values
+	indexers := getTargetIndexers(metrics, o)
+	if len(indexers) == 0 {
+		return fmt.Errorf("failed to instantiate OpenSearch bulkindexer")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(5000000000))
+	defer cancel()
+
+	for _, metric := range metrics {
+		var name = metric.Name()
+
+		// index name has to be re-evaluated each time for telegraf
+		// to send the metric to the correct time-based index
+		indexName := o.GetIndexName(o.IndexName, metric.Time(), o.tagKeys, metric.Tags())
+
+		// Handle NaN and inf field-values
+		fields := make(map[string]interface{})
+		for k, value := range metric.Fields() {
+			v, ok := value.(float64)
+			if !ok || o.FloatHandling == "none" || !(math.IsNaN(v) || math.IsInf(v, 0)) {
+				fields[k] = value
+				continue
+			}
+			if o.FloatHandling == "drop" {
+				continue
+			}
+
+			if math.IsNaN(v) || math.IsInf(v, 1) {
+				fields[k] = o.FloatReplacement
+			} else {
+				fields[k] = -o.FloatReplacement
+			}
+		}
+
+		m := make(map[string]interface{})
+
+		m["@timestamp"] = metric.Time()
+		m["measurement_name"] = name
+		m["tag"] = metric.Tags()
+		m[name] = fields
+
+		body, err := json.Marshal(m)
+		if err != nil {
+			return fmt.Errorf("failed to marshal body: %w", err)
+		}
+
+		bulkIndxrItem := opensearchutil.BulkIndexerItem{
+			Action:    "index",
+			Index:     indexName,
+			Body:      strings.NewReader(string(body)),
+			OnSuccess: o.onSucc,
+			OnFailure: o.onFail,
+		}
+		if o.ForceDocumentID {
+			bulkIndxrItem.DocumentID = getPointID(metric)
+		}
+
+		if o.UsePipeline != "" {
+			if pipelineName := o.getPipelineName(o.pipelineName, o.pipelineTagKeys, metric.Tags()); pipelineName != "" {
+				if indexers[pipelineName] != nil {
+					if err := indexers[pipelineName].Add(ctx, bulkIndxrItem); err != nil {
+						o.Log.Errorf("error adding metric entry to OpenSearch bulkIndexer: %w for pipeline %s", err, pipelineName)
+					}
+					continue
+				}
+			}
+		}
+
+		if err := indexers["default"].Add(ctx, bulkIndxrItem); err != nil {
+			o.Log.Errorf("error adding metric entry to OpenSearch default bulkIndexer: %w", err)
+		}
+	}
+
+	for _, bulkIndxr := range indexers {
+		if err := bulkIndxr.Close(ctx); err != nil {
+			return fmt.Errorf("error sending bulk request to OpenSearch: %w", err)
+		}
+
+		// Report the indexer statistics
+		stats := bulkIndxr.Stats()
+		if stats.NumAdded < uint64(len(metrics)) {
+			return fmt.Errorf("OpenSearch indexed [%d] documents with [%d] errors", stats.NumAdded, stats.NumFailed)
+		}
+		o.Log.Debugf("OpenSearch successfully indexed [%d] documents\n", stats.NumAdded)
```

Comment:
I suggest to remove the `OpenSearch` term from all error messages, here and elsewhere! This information is already encoded by the logger so no need to duplicate that here. Additionally this will solve the "errors have to start lowercase" issue...

Reply:
done